tests/versioning_test.go

```diff
@@ -180,72 +175,825 @@ func (s *versioningIntegSuite) TestVersioningStateNotDestroyedByOtherUpdates() {
+	s.addNewDefaultBuildId(ctx, tq, "v1")
+	s.waitForPropagation(ctx, tq, "v1")
+
+	// unblock the workflow
+	s.NoError(s.sdkClient.SignalWorkflow(ctx, run.GetID(), run.GetRunID(), "wait", nil))
+
+	var out string
+	s.NoError(run.Get(ctx, &out))
+	s.Equal("done!", out)
+}
+
+func (s *versioningIntegSuite) TestDispatchUpgradeStickyTimeout() {
+	s.testWithMatchingBehavior(func() { s.dispatchUpgrade(true) })
+}
+func (s *versioningIntegSuite) TestDispatchUpgradeStickyUnavailable() {
+	s.testWithMatchingBehavior(func() { s.dispatchUpgrade(false) })
+}
+
+func (s *versioningIntegSuite) dispatchUpgrade(letStickyWftTimeout bool) {
+	tq := s.randomizeStr(s.T().Name())
+
+	var started sync.WaitGroup
+	started.Add(1)
+
+	wf1 := func(ctx workflow.Context) (string, error) {
+		started.Done()
+		workflow.GetSignalChannel(ctx, "wait").Receive(ctx, nil)
+		return "done!", nil
+	}
+
+	wf2 := func(ctx workflow.Context) (string, error) {
+		workflow.GetSignalChannel(ctx, "wait").Receive(ctx, nil)
+		return "done from two!", nil
+	}
+
+	// TODO: reduce after fixing sticky
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	defer cancel()
+
+	s.addNewDefaultBuildId(ctx, tq, "v1")
+	s.waitForPropagation(ctx, tq, "v1")
+
+	w1 := worker.New(s.sdkClient, tq, worker.Options{
+		BuildID:                 "v1",
+		UseBuildIDForVersioning: true,
+	})
+	w1.RegisterWorkflowWithOptions(wf1, workflow.RegisterOptions{Name: "wf"})
+	s.NoError(w1.Start())
+	defer w1.Stop()
+
+	run, err := s.sdkClient.ExecuteWorkflow(ctx, sdkclient.StartWorkflowOptions{TaskQueue: tq}, "wf")
+	s.NoError(err)
+	started.Wait()
+
+	// Stop w1 to break stickiness
+	// TODO: this shouldn't be necessary, add behavior cases that disable stickiness
+	w1.Stop()
+
+	// two methods of breaking stickiness:
+	if letStickyWftTimeout {
+		// in this case we just start the new worker and kick the workflow immediately. the new
+		// wft will go to the sticky queue, be spooled, but eventually timeout and we'll get a
+		// new wft.
+	} else {
+		// in this case we sleep for more than stickyPollerUnavailableWindow. matching will
+		// return StickyWorkerUnavailable immediately after that.
+		time.Sleep(11 * time.Second)
+	}
+
+	// now add v2 as compatible so the next workflow task runs there
+	s.addCompatibleBuildId(ctx, tq, "v2", "v1", false)
+	s.waitForPropagation(ctx, tq, "v2")
+
+	w2 := worker.New(s.sdkClient, tq, worker.Options{
+		BuildID:                 "v2",
+		UseBuildIDForVersioning: true,
+	})
+	w2.RegisterWorkflowWithOptions(wf2, workflow.RegisterOptions{Name: "wf"})
+	s.NoError(w2.Start())
+	defer w2.Stop()
+
+	// unblock the workflow
+	s.NoError(s.sdkClient.SignalWorkflow(ctx, run.GetID(), run.GetRunID(), "wait", nil))
+
+	var out string
+	s.NoError(run.Get(ctx, &out))
+	s.Equal("done from two!", out)
+}
+
+func (s *versioningIntegSuite) TestDispatchActivity() {
+	s.testWithMatchingBehavior(s.dispatchActivity)
+}
+
+func (s *versioningIntegSuite) dispatchActivity() {
```

Comment:
I would also test that an activity is dispatched from a v1 worker is processed by a v1.1 worker if that version has been added while the workflow is running.

Reply:
yup, will do